src/components/implementation/no_interface/new_boot/ppos_booter.c

```diff
@@ -0,0 +1,323 @@
+		struct cobj_sect *sect;
+		vaddr_t dest_daddr;
+		char *lsrc, *dsrc;
+		int left, dest_doff;
+
+		sect       = cobj_sect_get(h, i);
+		/* virtual address in the destination address space */
+		dest_daddr = sect->vaddr;
+		/* where we're copying from in the cobj */
+		lsrc       = cobj_sect_contents(h, i);
+		/* how much is left to copy? */
+		left       = cobj_sect_size(h, i);
+
+		/* Initialize memory. */
+		if (!(sect->flags & COBJ_SECT_KMEM) &&
+		    (first_time || !(sect->flags & COBJ_SECT_INITONCE))) {
+			if (sect->flags & COBJ_SECT_ZEROS) {
+      				memset(start_addr + (dest_daddr - init_daddr), 0, left);
+			} else {
+				memcpy(start_addr + (dest_daddr - init_daddr), lsrc, left);
+			}
+		}
+
+		if (sect->flags & COBJ_SECT_CINFO) {
+			assert(left == PAGE_SIZE);
+			assert(comp_info == dest_daddr);
+			boot_process_cinfo(h, spdid, boot_spd_end(h), start_addr + (comp_info-init_daddr), comp_info);
+			ci = (struct cos_component_information*)(start_addr + (comp_info-init_daddr));
+			comp_cap_info[h->id].upcall_entry = ci->cos_upcall_entry;
+			
+			vaddr_t begin =start_addr + ((ci->cos_upcall_entry) - init_daddr); 
+		}
+	
+	}
+
+	return 0;
+}
+
+static int
+boot_comp_map(struct cobj_header *h, spdid_t spdid, vaddr_t comp_info, pgtblcap_t pt)
+{
+	boot_comp_map_memory(h, spdid, pt);
+	boot_comp_map_populate(h, spdid, comp_info, 1);
+	return 0;
+}
+
+static void
+boot_create_cap_system(void)
+{
+	unsigned int i;
+
+
+	for(i = 0; hs[i] != NULL; i++){
+		
+		struct cobj_header *h;
+		struct cobj_sect *sect;
+		captblcap_t ct;
+		pgtblcap_t pt;
+		compcap_t cc;
+		spdid_t spdid;
+		capid_t pte_cap;
+		vaddr_t ci = 0;
+		
+		h = hs[i];
+		spdid = h->id;
+		printc("Initializing Comp: %s\n", h->name);	
+		
+		sect = cobj_sect_get(h, 0);
+		comp_cap_info[spdid].addr_start = sect->vaddr;
+		
+		ct = cos_captbl_alloc(&boot_info);
+		assert(ct);	
+		pt = cos_pgtbl_alloc(&boot_info);
+		assert(pt);	
+		
+		/*The comp_cap is not alloc'd yet, due to hack for upcall_fn*/
+		cos_compinfo_init(&comp_cap_info[spdid].cos_compinfo, pt, ct, 0, 
+				  (vaddr_t)sect->vaddr, 4, &boot_info);
+		
+		if (boot_spd_symbs(h, spdid, &ci)) BUG();
+		if (boot_comp_map(h, spdid, ci, pt))   BUG();
+	        	
+		cc = cos_comp_alloc(&boot_info, ct, pt, (vaddr_t)comp_cap_info[spdid].upcall_entry);
+		assert(cc);	
+		comp_cap_info[spdid].cos_compinfo.comp_cap = cc;
+	
+		/* Create sinv capability from Userspace to Booter components */
+		/* Need id 1 because we need to wait for id 0 to finish initializing */
+		sinvcap_t sinv;
+
+		sinv = cos_sinv_alloc(&boot_info, boot_info.comp_cap, (vaddr_t)__inv_test_entry);
+		assert(sinv > 0);
+		printc("sinv: %d\n", sinv);
+
+		/* Copy into vm0 capability table at a known location */
+		cos_cap_cpy_at(&comp_cap_info[spdid].cos_compinfo, BOOT_SINV_CAP, &boot_info, sinv);
+
+		printc("Comp %d (%s) activated @ %x!\n", h->id, h->name, sect->vaddr);
+		
+		thdcap_t main_thd = cos_initthd_alloc(&boot_info, cc);
+		assert(main_thd);
```

Comment:
@ryuxin the idea is that the new componenet will sinv() back into the booter after it's done initing. @gparmer and I have discussed (and likely prefer) instead that the booter sinv()'s into the new components to be inited, instead of the other way around. this requires a bit more modification and is thus not done here.